utilities.c

```diff
@@ -0,0 +1,213 @@
+        if ((len = matches(buffer + index, "ISO MD5SUM = "))) {
+            index += len;
+            if (index + HASH_SIZE >= APPDATA_SIZE)
+                goto fail;
+
+            memcpy(result->hashsum, buffer + index, HASH_SIZE);
+            result->hashsum[HASH_SIZE] = '\0';
+            index += HASH_SIZE;
+            task |= TASK_MD5;
+            /* Skip to next semicolon. */
+            for (char *p = buffer + index; index < APPDATA_SIZE && *p != ';';
+                 p++, index++) {
+            }
+        } else if ((len = matches_number(buffer, index,
+                                         "SKIPSECTORS = ", (long int *) &result->skipsectors))) {
+            index = len;
+            if (index >= APPDATA_SIZE)
+                goto fail;
+            task |= TASK_SKIP;
+        } else if ((len = matches_number(buffer, index,
+                                         "RHLISOSTATUS=", (long int *) &result->supported))) {
+            index = len;
+            task |= TASK_SUPPORTED;
+        } else if ((len = matches(buffer + index, "FRAGMENT SUMS = "))) {
+            index += len;
+            if (index + FRAGMENT_SUM_SIZE >= APPDATA_SIZE)
+                goto fail;
+            memcpy(result->fragmentsums, buffer + index, FRAGMENT_SUM_SIZE);
+            result->fragmentsums[FRAGMENT_SUM_SIZE] = '\0';
+            task |= TASK_FRAGSUM;
+            index += FRAGMENT_SUM_SIZE;
+            for (char *p = buffer + index; index < APPDATA_SIZE && *p != ';';
+                 p++, index++) {
+            }
+        } else if ((len = matches_number(buffer, index, "FRAGMENT COUNT = ", (long int *) &result->fragmentcount))) {
+            index = len;
+            task |= TASK_FRAGCOUNT;
+        }
+        /* Either something is wrong or it skips a semicolon. */
+        index++;
+        if (task == TASK_DONE)
+            break;
+    }
+
+    if (task < TASK_SKIP + TASK_MD5) {
+    fail:
+        free(result);
+        return NULL;
+    }
+    return result;
+}
+
+bool validate_fragment(const MD5_CTX *const hashctx, const size_t fragment,
+                       const size_t fragmentsize, const char *const fragmentsums, char *hashsums) {
+    unsigned char digest[HASH_SIZE / 2];
+    MD5_CTX ctx;
+    memcpy(&ctx, hashctx, sizeof(ctx));
+    MD5_Final(digest, &ctx);
+    size_t j = (fragment - 1) * fragmentsize;
+    for (size_t i = 0; i < MIN(fragmentsize, HASH_SIZE / 2); i++) {
+        char tmp[2];
+        snprintf(tmp, 2, "%01x", digest[i]);
+        if (hashsums != NULL)
+            strncat(hashsums, tmp, 1);
+        /* printf("%c == %c\n", tmp[0], fragmentsums[j]); */
+        if (fragmentsums != NULL && tmp[0] != fragmentsums[j++])
+            return false;
+    }
+    return true;
+}
+
+void md5sum(char *const hashsum, MD5_CTX *const hashctx) {
+    unsigned char digest[HASH_SIZE / 2];
+    MD5_Final(digest, hashctx);
+    *hashsum = '\0';
+    for (size_t i = 0; i < HASH_SIZE / 2; i++) {
+        char tmp[3];
```

Comment:
@bcl regarding the `snprintf` comment above. Why is that? Should it be four here then because of `"%02x"`?